cartographer_ros/cartographer_ros/offline_node.cc

```diff
@@ -132,101 +153,139 @@ void RunOfflineNode(const MapBuilderFactory& map_builder_factory) {
-        auto tf_message = msg.instantiate<tf2_msgs::TFMessage>();
-        tf_publisher.publish(tf_message);
-
-        for (const auto& transform : tf_message->transforms) {
-          try {
-            tf_buffer.setTransform(transform, "unused_authority",
-                                   msg.getTopic() == kTfStaticTopic);
-          } catch (const tf2::TransformException& ex) {
-            LOG(WARNING) << ex.what();
-          }
-        }
+    } else {
+      for (const std::string& topic : cartographer_ros::SplitString(
+               specified_sensor_topics.at(current_bag_index), ':')) {
+        CHECK(current_expected_sensor_topics
+                  .insert(node.node_handle()->resolveName(topic))
+                  .second);
       }
+    }
+    CHECK(expected_sensor_topics
+              .emplace(std::piecewise_construct,
+                       std::forward_as_tuple(current_bag_index),
+                       std::forward_as_tuple(current_expected_sensor_topics))
+              .second);
 
-      while (!delayed_messages.empty() &&
-             delayed_messages.front().getTime() < msg.getTime() - kDelay) {
-        const rosbag::MessageInstance& delayed_msg = delayed_messages.front();
-        const std::string topic = node.node_handle()->resolveName(
-            delayed_msg.getTopic(), false /* resolve */);
-        if (delayed_msg.isType<sensor_msgs::LaserScan>()) {
-          node.HandleLaserScanMessage(
-              trajectory_id, topic,
-              delayed_msg.instantiate<sensor_msgs::LaserScan>());
-        }
-        if (delayed_msg.isType<sensor_msgs::MultiEchoLaserScan>()) {
-          node.HandleMultiEchoLaserScanMessage(
-              trajectory_id, topic,
-              delayed_msg.instantiate<sensor_msgs::MultiEchoLaserScan>());
-        }
-        if (delayed_msg.isType<sensor_msgs::PointCloud2>()) {
-          node.HandlePointCloud2Message(
-              trajectory_id, topic,
-              delayed_msg.instantiate<sensor_msgs::PointCloud2>());
-        }
-        if (delayed_msg.isType<sensor_msgs::Imu>()) {
-          node.HandleImuMessage(trajectory_id, topic,
-                                delayed_msg.instantiate<sensor_msgs::Imu>());
-        }
-        if (delayed_msg.isType<nav_msgs::Odometry>()) {
-          node.HandleOdometryMessage(
-              trajectory_id, topic,
-              delayed_msg.instantiate<nav_msgs::Odometry>());
-        }
-        if (delayed_msg.isType<sensor_msgs::NavSatFix>()) {
-          node.HandleNavSatFixMessage(
-              trajectory_id, topic,
-              delayed_msg.instantiate<sensor_msgs::NavSatFix>());
-        }
-        clock.clock = delayed_msg.getTime();
-        clock_publisher.publish(clock);
+    playable_bag_multiplexer.AddPlayableBag(PlayableBag(
+        bag_filename, current_bag_index, ros::TIME_MIN, ros::TIME_MAX, kDelay,
+        [&](const rosbag::MessageInstance& msg) {
```

Comment:
Please explicitly write down the captures (and prefer copies if possible), I cannot see what is captured.

Reply:
`tf_publisher` and `tf_buffer` are captured. The whole point of the lambda is not to do filtering (which I thought just looked nice), but rather to insert the transform into `tf_buffer`. This was previously achieved by having `delayed_messages`.